### PR TITLE
gh-110944: Make pdb completion work for alias and convenience vars

### PR DESCRIPTION
**Pull request by gaogaotiantian**
`$` was removed from the delimiters as it was used for convenience variables.

* Issue: gh-110944

### CODE REVIEW

File: Lib/pdb.py [hunk 1]
```diff
@@ -232,7 +232,7 @@ def __init__(self, completekey='tab', stdin=None, stdout=None, skip=None,
         try:
             import readline
             # remove some common file name delimiters
-            readline.set_completer_delims(' \t\n`@#$%^&*()=+[{]}\\|;:\'",<>?')
+            readline.set_completer_delims(' \t\n`@#%^&*()=+[{]}\\|;:\'",<>?')
```
Comment (iritkatriel) [inline Lib/pdb.py:241]: There are no tests in this PR. 

What could this change break?

Comment (gaogaotiantian) [reply Lib/pdb.py:241]: There's no test for completion, presumably because it's not easy to test. It would change the behavior when the user tries to do a complete (hit `<tab>` in most cases) and there's a `$` in the line. In that case, `$` would be not considered as a delimiter (the text to be completed would include `$`). I can't think of any real cases where this matters. Actually, I can't think of a case where the user uses `$` in their pdb commands (other than using convenience variable of course).

File: Lib/pdb.py [hunk 2]
```diff
@@ -670,6 +670,19 @@ def set_convenience_variable(self, frame, name, value):
     # Generic completion functions.  Individual complete_foo methods can be
     # assigned below to one of these functions.
 
+    def completenames(self, text, line, begidx, endidx):
```
Comment (iritkatriel) [inline Lib/pdb.py:689]: Is this used anywhere?

Comment (gaogaotiantian) [reply Lib/pdb.py:689]: This is overwritting the `completenames` method defined in `cmd.Cmd`, which is called when the user tries to complete the command.

`Pdb` class has some implicit inheritance from `cmd.Cmd` class to achieve customized behaviors. I can add some comments here if it helps.

Comment (iritkatriel) [reply Lib/pdb.py:689]: I think adding a test that covers this would be better.

Comment (gaogaotiantian) [reply Lib/pdb.py:689]: I don't think you can (easily) set up a readline test to really trigger the "complete" feature of readline. Even `test_cmd` does not test against complete features for `cmd.Cmd`. I can add some unitttest just to test this method (feeding mock arguments to the function), but a real functional test involving readline would be a large amount of work - and we should probably do that for `cmd` first if we consider adding it.

**Comment by encukou** [comment]
Would you be willing to add that test? It would make it much easier to review this change.

There is an existing `pty`-based helper in `test_readline` that could be moved to a shared `test.support.pty_helper.py` file for reuse in `pdb` tests: https://github.com/python/cpython/blob/main/Lib/test/test_readline.py#L307

**Comment by gaogaotiantian** [comment]
> Would you be willing to add that test? It would make it much easier to review this change.
> 
> There is an existing `pty`-based helper in `test_readline` that could be moved to a shared `test.support.pty_helper.py` file for reuse in `pdb` tests: https://github.com/python/cpython/blob/main/Lib/test/test_readline.py#L307

I put the helper into a new file as you said, then added some completion tests for pdb. It's not "complete", but it's at least something, and it tests the changes I made for this PR. I'll see if there are more interesting cases now that we have the structure.

**Comment by gaogaotiantian** [comment]
Hi @iritkatriel , tests were added for both pdb and cmd. After we merge this, we can do the cmd libedit support. (Or we can do it vice versa, but writing test for the other PR is difficult).

This PR:
* Disables a test on libedit
* Introduces the helper for pty

The other PR:
* Supports libedit for cmd
* Needs the pty helper to write tests

So either
* This PR -> the other PR with tests -> Remove the disable of this PR, or
* The other PR without tests -> This PR with removed disable check -> tests to the other PR.

**Comment by iritkatriel** [comment]
I'm not familiar with libedit/pty. I think @serhiy-storchaka might be?

File: Lib/pdb.py [hunk 3]
```diff
@@ -686,6 +686,19 @@ def set_convenience_variable(self, frame, name, value):
...
+        # Overwrite completenames() of cmd so for the command completion,
+        # if no current command matches, check for expressions as well
+        dotext = 'do_' + text
+        commands = [a[3:] for a in self.get_names() if a.startswith(dotext)]
```
Comment (encukou) [inline Lib/pdb.py]: Use the superclass implementation, don't copy it.
```suggestion
        commands = super().completenames(text, line, begidx, endidx)
```

File: Lib/test/test_pdb.py [hunk 4]
```diff
@@ -3235,6 +3237,47 @@ def test_checkline_is_not_executable(self):
...
+        if readline.__doc__ and "libedit" in readline.__doc__:
+            raise unittest.SkipTest("libedit readline is not supported for pdb")
+
+    def test_basic_completion(self):
+        script = textwrap.dedent("""
+            import pdb; pdb.Pdb().set_trace()
+            print('hello')
+        """)
+
+        input = b"co\t\tntin\t\n"
```
Comment (encukou) [inline Lib/test/test_pdb.py]: This is hard to follow, I suggest adding a comment.
```suggestion
        # List everything starting with 'co',
        # then complete and enter 'continue'
        input = b"co\t\tntin\t\n"
```

File: Lib/test/test_pdb.py [hunk 5]
```diff
@@ -3235,6 +3237,47 @@ def test_checkline_is_not_executable(self):
...
+    def test_expression_completion(self):
+        script = textwrap.dedent("""
+            value = "speci"
+            import pdb; pdb.Pdb().set_trace()
+        """)
+
+        input = b"val\t + 'al'\n"
+        input += b"p val\t + 'es'\n"
+        input += b"$_fra\t\n"
+        input += b"c\n"
```
Comment (encukou) [inline Lib/test/test_pdb.py]: Here too.
```suggestion
        # Complete: value + 'al'
        input = b"val\t + 'al'\n"
        # Complete: p value + 'es'
        input += b"p val\t + 'es'\n"
        # Complete: $_frame
        input += b"$_fra\t\n"
        # Continue
        input += b"c\n"
```

**Review by encukou — COMMENTED** [review]
Thank you. I'm not very familiar with libedit/pty either, but this looks reasonable to me. If we don't find a better expert, let's merge it!

It would be best if you could split this PR into one that adds the `test_basic_completion` test, and the one that adds the functionality (and the test for it).
The basic test could then be more easily backported to maintenance versions of Python. It's OK to link both PRs to the same issue.

I also have a few nitpicks below.

(If you don't want to take on the extra work, let me know -- I can take over)

**Comment by gaogaotiantian** [comment]
@encukou sure I cherry-picked the pty-helper piece and the basic test to another PR https://github.com/python/cpython/pull/111826. Let's finish that PR and circle back to this one.

**Comment by gaogaotiantian** [comment]
@encukou I've applied the suggestions on the code left :) Let me know if there's anything missing from the PR.

**Comment by encukou** [comment]
Looks good, thank you!
